### ISSUE TITLE
coluna initialization with generated solution

### ISSUE THREAD

**Issue by ysu1028** [atoptima/Coluna.jl#614]
Hello,

I'm wondering if I can feed Coluna with the initial solution that I generated from my meta-heuristic. I noticed that in the first iteration of Coluna, the dual bound is very bad (like -1500000, you can see in the 
<img width="422" alt="coluna" src="https://user-images.githubusercontent.com/58483379/141282645-1a540b8d-2ec3-4756-b742-bd073b2b93f8.png">
attached figure) and is gradually improved. Therefore, it takes a lot of time to reach a good dual bound value if we run Coluna directly. However, I have very good solutions at hand and I would like to know if it is possible to input my solution into Coluna so that Coluna can start with a very good dual bound?

I've searched a lot, it seems that I should do: set_ip_primal_sol (or add_ip_primal_sol ?) before solving. And defining an OptimizationState to call these functions. (link below)

https://docs.juliahub.com/Coluna/fKGQ4/0.3.8/dev/todo/#Coluna.ColunaBase.Solution

But I'm not clear on how to define a solution in the type of Coluna and I don't find any other documents about this issue.

Could you give me an example of solution format or more information on the Coluna initialization?

Thank you very much! I'm very looking forward to your reply!

Best,
Yue

**Comment by guimarqu** [comment]
Hi,

You can't easily do such an initialization at the moment (see #514) but this is, for sure, a must-have feature.

If you want to try the hard way (on the master branch commit `7679fcbae5`) : 

1- you must add the subproblems solution before starting the algorithm. You can add a callback here : https://github.com/atoptima/Coluna.jl/blob/7679fcbae5767be79d4945b78da71bf76202a63b/src/optimize.jl#L85

something like : 
```julia
mycallback(reform, env)
```

2- You can use the following method in your callback to add columns to the subproblems.

https://github.com/atoptima/Coluna.jl/blob/7679fcbae5767be79d4945b78da71bf76202a63b/src/MathProg/formulation.jl#L334-L342 

See this example : 

https://github.com/atoptima/Coluna.jl/blob/7679fcbae5767be79d4945b78da71bf76202a63b/test/node_finalizer_tests.jl#L126-L132



It will looks something like : 

```julia

jump_model = nothing # we need the JuMP model in the global scope (we need a UI to avoid this)

function my_jump_model()
  model = Model()
  # JuMP model definition here
  jump_model = model
end

function my_callback(reform, env)
  # Build solutions for each subproblems

  # You have to look for the correct subproblem in get_dw_pricing_sps(reform)
  # it returns a dict ( FormId => Formulation ), be careful, FormId != AxisId.
  # If you know to which subproblem each variable belong, you can try to look for this variable in each Formulation

  # You can have access to the JuMP variable through jump_model[:x] for instance
  x = jump_model[:x]
  nz_vars = [x[1,1], x[1,3]]
  nz_vals = [1.0, 4.0]

  # create the Primal Solution
  nz_varids = Coluna._get_orig_varid_in_form(JuMP.backend(jump_model), spform, v) for v in JuMP.index.(nz_vars)

  # don't forget the setup var in the solution
  push!(nz_varids, spform.duty_data.setup_var) 
  push!(nz_vals, 1.0) # multiplicity of the column

 sol = Coluna.MathProg.PrimalSolution(spform, nz_varids, nz_vals, sp_sol_cost, Coluna.FEASIBLE_SOL) 
 _, col_id = ClMP.insert_column!(masterform, sol, "MC")  # column is inserted !

end

```

### Refs :

https://github.com/atoptima/Coluna.jl/blob/7679fcbae5767be79d4945b78da71bf76202a63b/src/MOIwrapper.jl#L113

https://github.com/atoptima/Coluna.jl/blob/7679fcbae5767be79d4945b78da71bf76202a63b/src/MathProg/reformulation.jl#L52

https://github.com/atoptima/Coluna.jl/blob/7679fcbae5767be79d4945b78da71bf76202a63b/src/MathProg/reformulation.jl#L77

https://github.com/atoptima/Coluna.jl/blob/f68c49ce73bad7a542389c0f7704eb41bf316d08/src/MathProg/solutions.jl#L13

**Comment by guimarqu** [comment]
You can try to use stabilization in column generation to improve convergence : 

https://atoptima.github.io/Coluna.jl/stable/man/algorithm/#Coluna.Algorithm.ColumnGeneration

**Comment by guimarqu** [comment]
closed by #656
